File: tests/integration/discs/test_break.c
/*
 * test_break.c - tests the BBC BRK handler via set_brk_ret / disarm_brk_ret.
 *
 * set_brk_ret() has setjmp/longjmp semantics:
 *   - it returns 0 the first time (the handler is now "armed");
 *   - when a subsequent non-ESC BRK fires, the handler long-jumps back to the
 *     set_brk_ret() call site, which then "returns" a second time with 1.
 *
 * cause_brk_func() (brk_helper.s) executes a BRK with a non-ESC error number,
 * so the expected sequence is:
 *   ARMED    (first return, r == 0)
 *   CAUGHT   (second return via the BRK long-jump, r == 1)
 *
 * If the handler did not catch the BRK the language's own BRK handler would
 * abort the program with an error message and neither CAUGHT nor the trailing
 * marker would appear.
 */
#include <conio.h>

extern unsigned char __fastcall__ set_brk_ret(void);
extern void          __fastcall__ disarm_brk_ret(void);
extern void          cause_brk_func(void);

int main(void) {
    unsigned char r;

    clrscr();

    r = set_brk_ret();
    if (r == 0) {
        cputs("ARMED\r\n");
        cause_brk_func();      /* BRK -> long-jumps back into set_brk_ret() */
        cputs("NOTRAP\r\n");   /* should never be reached */
        return 1;
    }

    /* Second return: we arrived here via the BRK long-jump (r == 1). */
    cputs("CAUGHT\r\n");
    disarm_brk_ret();
    cputs("DONE\r\n");
    return 0;
}

File: tests/integration/scripted/test_break.py
"""Integration test: BBC BRK handler (set_brk_ret / disarm_brk_ret).

Runs discs/test_break.c (built as TBREAK). set_brk_ret() has setjmp/longjmp
semantics: it returns 0 once (armed), and a subsequent non-ESC BRK long-jumps
back so it "returns" a second time with 1. The program prints ARMED, triggers a
BRK, and on the second return prints CAUGHT then DONE. The "NOTRAP" marker must
never appear (it would mean the BRK was not caught).
"""

from __future__ import annotations

import pytest

import clib_harness as h


def test_brk_handler_catches_break(bbc_for_mode, mode):
    h.reset_and_boot(bbc_for_mode, "TBREAK", mode)
    h.run_disc_program(bbc_for_mode, "TBREAK")

    h.assert_screen(bbc_for_mode, "ARMED")
    h.assert_screen(bbc_for_mode, "CAUGHT")
    h.assert_screen(bbc_for_mode, "DONE")
    # If the handler had not trapped the BRK, the program would have fallen
    # through to NOTRAP (or the language would have aborted it).
    h.assert_not_screen(bbc_for_mode, "NOTRAP")
